envoy/server/transport_socket_config.h

```diff
@@ -30,14 +30,23 @@ class TransportSocketFactoryContext {
   virtual ~TransportSocketFactoryContext() = default;
 
   /**
-   * @return ServerFactoryContext which lifetime is no shorter than the server.
+   * @return ServerFactoryContext& the server factory context.
    */
-  virtual ServerFactoryContext& getServerFactoryContext() PURE;
+  virtual ServerFactoryContext& serverFactoryContext() PURE;
 
   /**
-   * @return OptRef<Server::Admin> the global HTTP admin endpoint for the server.
+   * @return Upstream::ClusterManager& singleton for use by the entire server.
+   * TODO(wbpcode): clusterManager() of ServerFactoryContext still be invalid when loading
+   * static cluster. So we need to provide an cluster manager reference here.
+   * This could be removed after https://github.com/envoyproxy/envoy/issues/26653 is resolved.
    */
-  virtual OptRef<Server::Admin> admin() PURE;
+  virtual Upstream::ClusterManager& clusterManager() PURE;
+
+  /**
+   * @return ProtobufMessage::ValidationVisitor& validation visitor for cluster configuration
+   * messages.
+   */
+  virtual ProtobufMessage::ValidationVisitor& messageValidationVisitor() PURE;
```

Comment:
This method is available in the `ServerFactoryContext&`. Can you clarify the reason for adding it in this interface, polease?

Reply:
The visitor of server factory context will be static vistor or dynamic vistor based on if the server is initialized.

But in the upstream cluster, the `added_via_api_` will be used as flag for choosing static vistor or dynamic visitor. The upstream transport factory context will use the clusters' visitor and it may be different with the visitor of server factory context.